### PR TITLE
[DO NOT MERGE] Automotive Skill - VehicleSettings support (english only)

### PR DESCRIPTION
**Pull request by darrenj**
## Description
A new Automotive Skill initially with VehicleSettings support. Currently only supports English but additional language support will be added early in Vanadium. The Skill unlike others is a partial implementation as every customer will have to integrate into their car infrastructure, for testing purposes we use Events to surface changes to the calling app/device.

## Related Issue
#59 

## Testing Steps

Unit Tests are provided for core scenarios. Test the Skill in "non skill-mode" directly as a Bot in the emulator. Due to the large LUIS models there isn't dispatch support at this time but will be integrated into VA early in Vanadium.

### Climate Control 
- *Set temperature to 21 degrees* 
- *Defog my windshield* 
- *Put the air on my feet* 
- *Turn off the ac* 
- *I'm feeling cold* 
- *It's feeling cold in the back* 
- *The passenger is freezing* 
- *Change climate control* 

### Safety 
- *Turn lane assist off* 
- *Enable lane change alert* 
- *Set park assist to alert* 

### Audio 
- *Adjust the equalizer* 
- *Increase the bass* 
- *Increase the volume*

## Checklist
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the appropriate unit tests
- [x] I have updated related documentation

If this contains changes that needs to be replicated between the Enterprise Template <-> Virtual Assistant
- [ ] A duplicate issue is filed to track future  work.

### CODE REVIEW

File: solutions/Virtual-Assistant/docs/virtualassistant-skills-automotive.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+- VEHICLE_SETTINGS_CHANGE
+- VEHICLE_SETTINGS_DECLARATIVE
+
+In addition there are two supporting LUIS models `settings_name` and `settings_value`, these are used for disambiguation scenarios to clarify setting names and values where the initial utterance doesn't provide clear information.
+
+## Customising Vehicle Settings
+
+Available vehicle settings are defined in a supporting metadata file which you can find in this location:  `automotiveskill\Dialogs\VehicleSettings\Resources\available_settings.json`.
+
+To add an new setting along with appropriate setting values it's easily expressed in JSON. The example below shows a new Volume control setting with the ability to Set, Increase, Decrease and Mute the volume.
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/docs/virtualassistant-skills-automotive.md:86]: To add 'a' new setting :)

File: solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln [hunk 2]
```diff
@@ -33,10 +33,14 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Tests", "Tests", "{DFCBFFD7
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Bot.Solutions.Tests", "tests\Microsoft.Bot.Solutions.Tests\Microsoft.Bot.Solutions.Tests.csproj", "{18366252-CEA5-4FB9-A022-ED957B491237}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CalendarSkillTest", "skills\tests\calendarskilltest\CalendarSkillTest.csproj", "{F67AE5B4-DC61-4C96-947B-87D054191734}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CalendarSkillTest", "skills\tests\calendarskilltest\CalendarSkillTest.csproj", "{F67AE5B4-DC61-4C96-947B-87D054191734}"
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln:36]: this change is probably unintentional right?

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj [hunk 3]
```diff
@@ -0,0 +1,214 @@
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <UserSecretsId>6a3184c3-074e-45b9-ad93-eceb8268ec01</UserSecretsId>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <ApplicationInsightsResourceId>/subscriptions/4c0d07e4-5490-4784-8d89-d8c504ba2a75/resourcegroups/LamillsBot/providers/microsoft.insights/components/LamillsBot-insights</ApplicationInsightsResourceId>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj]: do you need to explicitly have these?

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/Chunk.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace AutomotiveSkill.Common.Number
+{
+    /// <summary>
+    /// A chunk of a string, which can be either a number or a non-number
+    /// substring between two numbers. Iff it is a number, its numeric value is
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/Chunk.cs]: Iff -> If

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+            // The Setting entity will contain any identified vehicle setting that was present in the utterance, e.g. front right airflow
+            // The Value entity will contain any identified value relating to a vehicle setting that was present in the utterance, e.g. warm
+            IList<AvailableSetting> selected_settings = new List<AvailableSetting>();
+
+            if (state.Entities.ContainsKey("SETTING"))
+            {
+                // If we have a Setting then try to find a match between the setting name provided and the available settings
+                selected_settings = this.settingMatcher.MatchSettingNamesExactly(state.Entities["SETTING"].First());
+
+                // If we have not found a exact setting match but we have a value then combine Setting and Value together to identify a match
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs]: an exact setting match

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+        {
+            IList<SettingMatch> setting_matches = new List<SettingMatch>();
+            var has_matching_value_for_any_setting = false;
+            ISet<string> setting_names_to_remove = new HashSet<string>();
+
+            // The Setting entity will contain any identified vehicle setting that was present in the utterance, e.g. front right airflow
+            // The Value entity will contain any identified value relating to a vehicle setting that was present in the utterance, e.g. warm
+            IList<AvailableSetting> selected_settings = new List<AvailableSetting>();
+
+            if (state.Entities.ContainsKey("SETTING"))
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs:69]: check the variable state as well as state.Entities before using them

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+                            }
+
+                            if (!Util.IsNullOrEmpty(validReplacements))
+                            {
+                                state.Entities.Remove("VALUE");
+                                foreach (var replacement in validReplacements)
+                                {
+                                    validChanges.Add(replacement);
+                                }
+                            }
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs:315]: all the 
else
{
    invalidChanges.Add(change);
}

code can be folded together if you add 

break;

in the if condition. then you can just add 
invalidChanges.Add(change);

after outside of the outermost if statement

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 8]
```diff
@@ -0,0 +1,734 @@
...
+
+        /// <summary>
+        /// Apply the selecting setting value to the setting values.
+        /// </summary>
+        /// <param name="state">State object.</param>
+        /// <param name="entityValues">List of entity values.</param>
+        /// <returns>Setting.</returns>
+        public IList<SettingChange> ApplySelectionToSettingValues(AutomotiveSkillState state, List<string> entityValues)
+        {
+            var settingValues = state.GetUniqueSettingValues();
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs:361]: check the parameter state and entityValues before using them

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 9]
```diff
@@ -0,0 +1,734 @@
...
+        /// <summary>
+        /// Validate a change.
+        /// </summary>
+        /// <param name="setting">A setting for validation.</param>
+        /// <returns>Validation Status enumeration.</returns>
+        private ValidationStatus ValidateChange(SettingChange setting)
+        {
+            ValidationStatus validity = ValidationStatus.Valid;
+
+            if (string.IsNullOrEmpty(setting.SettingName))
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs:431]: if setting != null &&

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs [hunk 10]
```diff
@@ -0,0 +1,734 @@
...
+                foreach (var selected_value in selected_values)
+                {
+                    selectedSettingValues.Add(selected_value.Value.CanonicalName);
+                }
+            }
+
+            IList<SettingChange> newCandidates = new List<SettingChange>();
+            foreach (var candidate in state.Changes)
+            {
+                if (candidate == null)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs]: is it possible? what type is state.Changes?

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs (hunk 9 above)
Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingFilter.cs:431]: put this check before all other code

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingList.cs [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SettingList"/> class.
+        /// Provides access to the available settings that are stored in the available_settings JSON file.
+        /// </summary>
+        /// <param name="availableSettingsPath">Path to available settings file.</param>
+        /// <param name="settingAlternativeNamesPath">Path to alternative names file.</param>
+        public SettingList(string availableSettingsPath, string settingAlternativeNamesPath)
+        {
+            using (StreamReader reader = new StreamReader(availableSettingsPath))
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Common/SettingList.cs:42]: check the arguments before using them

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/Main/Resources/AutomotiveSkillMainResponses.tt [hunk 12]
```diff
@@ -0,0 +1,108 @@
+﻿<#@ assembly name="Newtonsoft.Json.dll" #>
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/Main/Resources/AutomotiveSkillMainResponses.tt]: There's a template we use for other skills for the templating file:
<#@ template debug="false" hostspecific="true" language="C#" #>
<#@ output extension=".cs" #>
<#@ include file="..\..\..\..\..\Microsoft.Bot.Solutions\Resources\ResponsesTemplate.t4"#>

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/VehicleSettings/VehicleSettingsDialog.cs [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+            { "increase", "Increasing" },
+        };
+
+        private readonly IRecognizer vehicleSettingNameSelectionLuisRecognizer;
+        private readonly IRecognizer vehicleSettingValueSelectionLuisRecognizer;
+
+        private readonly SettingList settingList;
+        private readonly SettingFilter settingFilter;
+
+        private string automotiveCarSettingImage = "https://msp2l1160225102310.blob.core.windows.net/ms-p2-l1-160225-1023-13-assets/Windows_Edu_inTune_539_v8_ContentPlacement-Choose3_img.jpg";
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/VehicleSettings/VehicleSettingsDialog.cs]: should we add this image file to the code?

**Review by lzc850612 — DISMISSED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj (hunk 3 above)
Comment (lauren-mills) [reply solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj]: No they need to be removed! A relic from the old telemetry implementation

File: solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln (hunk 2 above)
Comment (darrenj) [reply solutions/Virtual-Assistant/src/csharp/VirtualAssistant.sln:36]: I had to regress SLN changes due to a merge that went wrong, hence the GUID above ended up being roll, this was going back to previous value so OK. Have double checked.

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj (hunk 3 above)
Comment (darrenj) [reply solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/AutomotiveSkill.csproj]: Indeed - made a note to remove from skill template

File: solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/VehicleSettings/VehicleSettingsDialog.cs (hunk 13 above)
Comment (darrenj) [reply solutions/Virtual-Assistant/src/csharp/skills/automotiveskill/Dialogs/VehicleSettings/VehicleSettingsDialog.cs]: Good spot, meant to leverage HttpContext and didn't get back around to it. Done

**Pull request merged**